% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{book}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\usepackage{booktabs}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage[]{natbib}
\bibliographystyle{plainnat}
\usepackage{bookmark}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={Spatial Transcriptomics Self-Training},
  pdfauthor={Monash Genomics and Bioinformatics Platform (MGBP)},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{Spatial Transcriptomics Self-Training}
\author{Monash Genomics and Bioinformatics Platform (MGBP)}
\date{Compiled: May 02, 2025}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{1}
\tableofcontents
}
\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}

\chapter{Getting started}\label{getting-started}

Data:
\href{https://www.10xgenomics.com/datasets/ffpe-human-colorectal-cancer-data-with-human-immuno-oncology-profiling-panel-and-custom-add-on-1-standard}{Xenium Data}

\section{Summary}\label{summary}

This workshop, conducted by the Monash Genomics and Bioinformatics Platform, will cover Spatial Tanscriptomics Analysis

Important links:

\begin{itemize}
\tightlist
\item
  \href{set-up.html}{Installation and Setup instructions}
\item
  \href{../slides}{Slideshow introduction}
\item
  \href{solutions.html}{Challenge solutions} (no peeking!)
\end{itemize}

\chapter{Introduction}\label{introduction}

\section{What is Spatial Trancriptomics?}\label{what-is-spatial-trancriptomics}

\section{What Research questions can answer?}\label{what-research-questions-can-answer}

\chapter{Xenium}\label{xenium}

\subsection{Technology}\label{technology}

\subsection{Data}\label{data}

\chapter{COSmix}\label{cosmix}

\section{Technology}\label{technology-1}

\section{Data}\label{data-1}

\chapter{MerScope}\label{merscope}

\section{Technology}\label{technology-2}

\section{Data}\label{data-2}

\chapter{Preprocessing}\label{preprocessing}

\chapter{Xenium}\label{xenium-1}

\chapter{Cosmix}\label{cosmix-1}

\chapter{MerScope}\label{merscope-1}

\chapter{Normalisation}\label{normalisation}

\chapter{Xenium}\label{xenium-2}

\chapter{Cosmix}\label{cosmix-2}

\chapter{MerScope}\label{merscope-2}

\chapter{Dimensionality Reduction}\label{dimensionality-reduction}

\chapter{Xenium}\label{xenium-3}

\chapter{Cosmix}\label{cosmix-3}

\chapter{MerScope}\label{merscope-3}

\chapter{Clustering}\label{clustering}

\chapter{Xenium}\label{xenium-4}

\chapter{Cosmix}\label{cosmix-4}

\chapter{MerScope}\label{merscope-4}

\chapter{Cell Annotation}\label{cell-annotation}

\chapter{Xenium}\label{xenium-5}

\chapter{Cosmix}\label{cosmix-5}

\chapter{Cell Annotation}\label{cell-annotation-1}

\chapter{Integration}\label{integration}

\chapter{Cell to Cell Signaling}\label{cell-to-cell-signaling}

  \bibliography{book.bib,packages.bib}

\end{document}
